dvc/scm.py

```diff
@@ -0,0 +1,105 @@
+import os
+import git
+
+from dvc.exceptions import DvcException
+
+
+class FileNotInRepoError(DvcException):
+    pass
+
+#class Brancher(object):
+#    def __init__(self, scm, branch, create=False):
+#        self.scm = scm
+#        self.branch = branch
+#        self.create = create
+#
+#    def __enter__(self):
+#        if not self.branch:
+#            return
+#
+#        self.saved_branch = self.scm.branch()
+#        if not self.create:
+#            self.scm.checkout(self.branch)
+#        else:
+#            self.scm.branch(self.branch)
+#
+#    def __exit__(self, type, value, traceback):
+#        if not self.branch or not self.saved_branch:
+#            return
+#
+#        self.scm.checkout(saved_branch)
+
+
+class Base(object):
+    @staticmethod
+    def is_repo(root_dir):
+        return True
+
+    def ignore(self, path):
+        pass
+
+    def ignore_list(self, p_list):
+        return [self.ignore(path) for path in p_list]
+
+    def add(self, path):
+        pass
+
+    def add_list(self, p_list):
+        return [self.add(path) for path in p_list]
+
+    def commit(self, msg):
+        pass
+
+    def checkout(self, branch):
+        pass
+
+    def branch(self, branch):
+        pass
+
+    def brancher(self, branch, create=False):
+        return Brancher(self, branch, create)
+
+
+class Git(Base):
+    GITIGNORE = '.gitignore'
+    GIT_DIR = '.git'
+
+    def __init__(self, root_dir):
+        self.root_dir = root_dir
+        self.repo = git.Repo(root_dir)
+
+    @staticmethod
+    def is_repo(root_dir):
+        git_dir = os.path.join(root_dir, Git.GIT_DIR)
+        return os.path.isdir(git_dir)
+
+    def ignore(self, path):
+        entry = os.path.basename(path)
+        gitignore = os.path.join(os.path.dirname(path), self.GITIGNORE)
+
+        if not gitignore.startswith(self.root_dir):
+            raise FileNotInRepoError()
+
+        if os.path.exists(gitignore) and entry in open(gitignore, 'r').read():
+            return
+
+        open(gitignore, 'a').write('\n' + entry)
+
+    def add(self, path):
+        self.repo.index.add(path)
+
+    def commit(self, msg):
+        self.repo.index.commit(msg)
+
+    def checkout(self, branch):
+        self.repo.git.checkout(branch)
+
+    def branch(self, branch):
+        self.repo.git.branch(branch)
+
+
+def SCM(root_dir):
```

Comment:
What does SCM stands for? :)

Reply:
We've discussed it IRL, but just for the record: SourceControlManagement. Just a generic name for git(and in the future other systems) driver switcher.